### PR TITLE
SantaCache: Add very basic distribution test

### PR DESCRIPTION
**Pull request by russellhancox**

### CODE REVIEW

File: Source/santa-driver/SantaCache.h [hunk 1]
```diff
@@ -186,6 +185,35 @@ template<typename KeyT, typename ValueT> class SantaCache {
...
 
+  /**
+    Fill in the per_bucket_counts array with the number of entries in each bucket.
+
+    The per_buckets_count array will contain the per-bucket counts, up to the number
+    in array_size. The start_bucket parameter will determine which bucket to start off
+    with and upon return will contain either 0 if no buckets are remaining or the next
+    bucket to begin with when called again.
+  */
+  void bucket_counts(uint16_t *per_bucket_counts, uint16_t *array_size, uint64_t *start_bucket) {
```
Comment (nguyen-phillip) [inline Source/santa-driver/SantaCache.h]: Any reason to not just return a vector<uint16_t> directly?

File: Tests/LogicTests/SantaCacheTest.mm [hunk 2]
```diff
@@ -81,6 +83,46 @@ - (void)testCacheResetAtLimit {
   XCTAssertEqual(sut.get(6), 42);
 }
 
+- (void)testDistribution {
+  const int bucket_ratio = 5;
+  auto sut = SantaCache<uint64_t, uint64_t>(5000, bucket_ratio);
+
+  // Fill the cache with random keys, all set to 1.
```
Comment (nguyen-phillip) [inline Tests/LogicTests/SantaCacheTest.mm]: It might be a good idea to have the test keys as a parameter.

File: Source/santa-driver/SantaCache.h [hunk 3]
```diff
@@ -186,6 +185,35 @@ template<typename KeyT, typename ValueT> class SantaCache {
...
+    uint64_t start = *start_bucket;
```
Comment (tburgin) [inline Source/santa-driver/SantaCache.h]: check to make sure start_bucket is not a nullptr

Comment (tburgin) [inline Source/santa-driver/SantaCache.h]: check to make sure start_bucket is not a nullptr

File: Source/santa-driver/SantaCache.h [hunk 4]
```diff
@@ -186,6 +185,35 @@ template<typename KeyT, typename ValueT> class SantaCache {
...
+    if (*start_bucket + *array_size > bucket_count_) *array_size = bucket_count_;
```
Comment (tburgin) [inline Source/santa-driver/SantaCache.h]: you already dereferenced start_bucket, why not use start?
check to make sure array_size is not a nullptr

File: Source/santa-driver/SantaCache.h [hunk 5]
```diff
@@ -186,6 +185,35 @@ template<typename KeyT, typename ValueT> class SantaCache {
...
+
+    for (uint16_t i = 0; i < *array_size; ++i) {
+      int16_t count = 0;
```
Comment (tburgin) [inline Source/santa-driver/SantaCache.h]: uint16_t ?

File: Source/santa-driver/SantaCache.h [hunk 6]
```diff
@@ -186,6 +185,35 @@ template<typename KeyT, typename ValueT> class SantaCache {
...
+      struct bucket *bucket = &buckets_[start + i];
+      lock(bucket);
+      struct entry *entry = (struct entry *)((uintptr_t)bucket->head - 1);
+      while (entry != nullptr) {
+        if (entry->key != 0) count++;
```
Comment (tburgin) [inline Source/santa-driver/SantaCache.h]: ++count; because you can

File: Source/santa-driver/SantaCache.h [hunk 7]
```diff
@@ -186,6 +185,35 @@ template<typename KeyT, typename ValueT> class SantaCache {
...
+        entry = entry->next;
+      }
+      unlock(bucket);
+      (*start_bucket)++;
```
Comment (tburgin) [inline Source/santa-driver/SantaCache.h]: nit: To save a couple start_bucket dereferences... 
Increment `start` on l202. You could get rid of this line.
On L214: if (start >= bucket_count_) *start_bucket = 0;

File: Source/santa-driver/SantaCache.h (hunk 1 above)
Comment (russellhancox) [reply Source/santa-driver/SantaCache.h]: No vector in the kernel, sadly.

File: Source/santa-driver/SantaCache.h (hunk 3 above)
Comment (russellhancox) [reply Source/santa-driver/SantaCache.h]: Done, plus all the others

File: Source/santa-driver/SantaCache.h (hunk 4 above)
Comment (russellhancox) [reply Source/santa-driver/SantaCache.h]: Done and done

File: Source/santa-driver/SantaCache.h (hunk 5 above)
Comment (russellhancox) [reply Source/santa-driver/SantaCache.h]: Done

File: Source/santa-driver/SantaCache.h (hunk 6 above)
Comment (russellhancox) [reply Source/santa-driver/SantaCache.h]: Done

File: Source/santa-driver/SantaCache.h (hunk 7 above)
Comment (russellhancox) [reply Source/santa-driver/SantaCache.h]: Done

File: Tests/LogicTests/SantaCacheTest.mm (hunk 2 above)
Comment (russellhancox) [reply Tests/LogicTests/SantaCacheTest.mm]: Done, sort of.

File: Source/santa-driver/SantaCache.h [hunk 8]
```diff
@@ -186,6 +185,38 @@ template<typename KeyT, typename ValueT> class SantaCache {
...
+  /**
+    Fill in the per_bucket_counts array with the number of entries in each bucket.
+
+    The per_buckets_count array will contain the per-bucket counts, up to the number
+    in array_size. The start_bucket parameter will determine which bucket to start off
+    with and upon return will contain either 0 if no buckets are remaining or the next
+    bucket to begin with when called again.
+  */
+  void bucket_counts(uint16_t *per_bucket_counts, uint16_t *array_size, uint64_t *start_bucket) {
+    if (per_bucket_counts == nullptr || array_size == nullptr || start_bucket == nullptr) return;
```
Comment (tburgin) [inline Source/santa-driver/SantaCache.h]: nit: maybe add a return type? -1 for this case 0 at the end?

**Pull request merged**